### PR TITLE
(IAC-718) end to end tests for function-based DSC Resources

### PR DESCRIPTION
**Pull request by DavidS**

### CODE REVIEW

File: build.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    It "is installable" {
+      pdk bundle exec puppet module install --verbose pkg/*.tar.gz
+      $LASTEXITCODE | Should -Be 0
+    }
+  }
+
+  Context "when passing in invalid values" {
+    It "reports the error" {
+      New-PuppetDscModule -PowerShellModuleName "____DoesNotExist____" -ErrorAction Stop | Should -Throw
```
Comment (michaeltlombardi) [inline build.Tests.ps1:43]: I _think_ the statement to the left of the pipe needs to be wrapped in `{}` to evaluate correctly without introducing an actual error in the test run.

File: build.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  It "has a REFERENCE.md" {
+    Test-Path "import/powershellget/REFERENCE.md" | Should -BeTrue
+  }
+
+  It "has a type generated" {
+    Test-Path "import/powershellget/lib/puppet/type/dsc_psmodule.rb" | Should -BeTrue
+  }
+
+  Context "inside the module" {
```
Comment (michaeltlombardi) [inline build.Tests.ps1:21]: For these, do we want to leverage the `Invoke-PDKCommand` private function (we can dotsource it into the current scope for use outside the module) to do better introspection/success checking?

File: build.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,60 @@
+$here = Split-Path -Parent $MyInvocation.MyCommand.Path
+$sut = (Split-Path -Leaf $MyInvocation.MyCommand.Path) -replace '\.Tests\.', '.'
+# . "$here\$sut"
+$script = "$here\$sut"
+
+& $script -ErrorAction Stop -PowerShellModuleName PowerShellGet -PowerShellModuleVersion "2.1.3"
+
+Describe $script {
+  It "creates a module" {
```
Comment (michaeltlombardi) [inline build.Tests.ps1:9]: For any of these, do we want to verify any contents or just existance?

File: build.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    }
+  }
+
+  Context "when passing in invalid values" {
+    It "reports the error" {
+      New-PuppetDscModule -PowerShellModuleName "____DoesNotExist____" -ErrorAction Stop | Should -Throw
+    }
+  }
+
+  Context "when a valid manifest causes a run-time error" {
```
Comment (michaeltlombardi) [inline build.Tests.ps1:47]: The shape of this context and the following ones makes sense to me!

File: build.Tests.ps1 (hunk 1 above)
Comment (DavidS) [reply build.Tests.ps1]: yeah, that worked

File: build.Tests.ps1 (hunk 2 above)
Comment (DavidS) [reply build.Tests.ps1:34]: done.

File: build.Tests.ps1 (hunk 3 above)
Comment (DavidS) [reply build.Tests.ps1:22]: validation of the files is done with `pdk validate` later. Not sure how brittle-vs-useful specific content checks would be.

**Pull request merged**